components/merkledb/src/proof_list_index/lazy.rs

```diff
@@ -0,0 +1,876 @@
+// Copyright 2019 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of a Merkelized version of an array list (Merkle tree).
+
+pub use super::proof::{ListProof, ListProofError};
+
+use byteorder::{LittleEndian, ReadBytesExt, WriteBytesExt};
+
+use std::{
+    io::{Read, Write},
+    marker::PhantomData,
+    mem::size_of,
+    ops::{Bound, RangeBounds},
+};
+
+use exonum_crypto::{Hash, HASH_SIZE};
+
+use super::{key::ProofListKey, proof::ProofOfAbsence};
+use crate::{
+    hash::HashTag,
+    views::{
+        AnyObject, BinaryAttribute, IndexAccess, IndexAddress, IndexBuilder, IndexState, IndexType,
+        Iter as ViewIter, View,
+    },
+    BinaryKey, BinaryValue, ObjectHash,
+};
+
+const LEN_SIZE: usize = size_of::<u64>();
+
+// TODO: Implement pop and truncate methods for Merkle tree. (ECR-173)
+
+/// A Merkelized version of an array list that provides proofs of existence for the list items.
+///
+/// `ProofListIndex` implements a Merkle tree, storing elements as leaves and using `u64` as
+/// an index. `ProofListIndex` requires that elements implement the [`BinaryValue`] trait.
+///
+/// [`BinaryValue`]: ../trait.BinaryValue.html
+#[derive(Debug)]
+pub struct LazyListIndex<T: IndexAccess, V> {
+    base: View<T>,
+    state: IndexState<T, ProofListState>,
+    _v: PhantomData<V>,
+}
+
+/// An iterator over the items of a `ProofListIndex`.
+///
+/// This struct is created by the [`iter`] or
+/// [`iter_from`] method on [`ProofListIndex`]. See its documentation for details.
+///
+/// [`iter`]: struct.ProofListIndex.html#method.iter
+/// [`iter_from`]: struct.ProofListIndex.html#method.iter_from
+/// [`ProofListIndex`]: struct.ProofListIndex.html
+#[derive(Debug)]
+pub struct LazyListIndexIter<'a, V> {
+    base_iter: ViewIter<'a, ProofListKey, V>,
+}
+
+impl<T, V> AnyObject<T> for LazyListIndex<T, V>
+where
+    T: IndexAccess,
+    V: BinaryValue + std::fmt::Debug,
+{
+    fn view(self) -> View<T> {
+        self.base
+    }
+
+    fn object_type(&self) -> IndexType {
+        IndexType::ProofList
+    }
+
+    fn metadata(&self) -> Vec<u8> {
+        self.state.metadata().to_bytes()
+    }
+}
+
+#[derive(Debug, Default, Copy, Clone)]
+struct ProofListState {
```

Comment:
It's a bit confusing that `LazyListIndex` has `ProofListState`.

Reply:
My bad, I should say about it in PR description. I plan to replace ProofListIndex with LazyListIndex if the latter shows the required performance. Currently, two indices are needed for benchmarking.